### ISSUE TITLE
Illegal memory access Segmentation fault (core dumped) in llama module test llama module 测试内存访问越界

### ISSUE THREAD

**Issue by parrotsky** [Tencent/ncnn#5155]
## error log | 日志或报错信息 | ログ
Segmentation fault (core dumped)  innerproduct error
## context | 编译/运行环境 | バックグラウンド
latest ncnn version, gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0 
## how to reproduce | 复现步骤 | 再現方法
1. Python script to generate pt file 
2.  
```
import torch
from transformers import LlamaConfig
from transformers.models.llama.modeling_llama import LlamaDecoderLayer
import time
config = LlamaConfig()
# print (config)

# Initialize a LlamaDecoderLayer
llama_decoder_layer = LlamaDecoderLayer(config)
# print(llama_decoder_layer)

dummy_input = torch.rand(1, 128, config.hidden_size)
# dummy_input = torch.rand(1, config.max_position_embeddings, config.hidden_size)
output = llama_decoder_layer(dummy_input)
device = "cpu"
llama_decoder_layer = llama_decoder_layer.to(device)
llama_decoder_layer.eval()
mod = torch.jit.trace(llama_decoder_layer, dummy_input)
mod.save(f"llama-module.pt")`
```
4.  转换模型 ./pnnx llama-module.pt inputshape=[1,128,4096]f32
5.  测试模型需要的cpp测试用例：
6. `
7. 
```
#include <iostream>
#include <string>
#include <cassert>

#include "net.h"
// #include <algorithm>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include "benchmark.h"
#include "cpu.h"
#include "gpu.h"
#include <stdio.h>
#include <vector>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>



int main(int argc, char** argv)
{

    ncnn::Mat in(1, 128, 4096);
    ncnn::Net net;

    // net.opt.blob_allocator = &g_blob_pool_allocator;
    // net.opt.workspace_allocator = &g_workspace_pool_allocator;
    net.opt.use_vulkan_compute = false;
    //net.opt.use_winograd_convolution = true;
    net.opt.use_sgemm_convolution = true;
    net.opt.use_int8_inference = false;
    net.opt.use_fp16_packed = false;
    net.opt.use_fp16_storage = false;
    net.opt.use_fp16_arithmetic = false;
    net.opt.use_int8_storage = false;
    net.opt.use_int8_arithmetic = false;
    net.opt.use_packing_layout = false;
    net.opt.use_shader_pack8 = false;
    net.opt.use_image_storage = false;
    net.opt.num_threads = 6;
    ncnn::set_omp_dynamic(0);
    ncnn::set_omp_num_threads(6);
   std::string modelbin = "llama-module.ncnn.bin";
    std::string modelparam = "llama-module.ncnn.param";

    net.load_param(modelparam.c_str());
    net.load_model(modelbin.c_str());
    ncnn::Mat out;
    int g_loop_count = 1;

    ncnn::Extractor ex = net.create_extractor();
    // ex.set_light_mode(true);
    // ex.set_num_threads(4);
    double time_min = DBL_MAX;
    double time_max = -DBL_MAX;
    double time_avg = 0;

    // auto start_time = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < g_loop_count; i++)
    {
        double start = ncnn::get_current_time();

        ncnn::Extractor ex = net.create_extractor();
        ex.input("in0", in);
        ex.extract("out0", out);

        double end = ncnn::get_current_time();
        double time = end - start;
        time_min = std::min(time_min, time);
        time_max = std::max(time_max, time);
        time_avg += time;

    }
    time_avg /= g_loop_count;
    fprintf(stderr, "min = %7.2f, max = %7.2f, avg = %7.2f\n", time_min, time_max, time_avg);
    fprintf(stderr, "Duration(ms) : %.2f\n",  time_avg); // Unit: ms

    return 0;
}
```
`

## more | 其他 | その他
经过valgrind分析出现如下错误：
`

> ==22717== Memcheck, a memory error detector
> ==22717== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
> ==22717== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
> ==22717== Command: ./test_llama_module
> ==22717== 
> ==22717== Thread 3:
> ==22717== Invalid read of size 16
> ==22717==    at 0x48D39D: ncnn::innerproduct_sse(ncnn::Mat const&, ncnn::Mat&, ncnn::Mat const&, ncnn::Mat const&, int, ncnn::Mat const&, ncnn::Option const&) [clone ._omp_fn.2] (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x487A86D: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
> ==22717==    by 0x48AA608: start_thread (pthread_create.c:477)
> ==22717==    by 0x4D5E132: clone (clone.S:95)
> ==22717==  Address 0xb523d080 is not stack'd, malloc'd or (recently) free'd
> ==22717== 
> ==22717== 
> ==22717== Process terminating with default action of signal 11 (SIGSEGV): dumping core
> ==22717==  Access not within mapped region at address 0xB523D080
> ==22717==    at 0x48D39D: ncnn::innerproduct_sse(ncnn::Mat const&, ncnn::Mat&, ncnn::Mat const&, ncnn::Mat const&, int, ncnn::Mat const&, ncnn::Option const&) [clone ._omp_fn.2] (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x487A86D: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
> ==22717==    by 0x48AA608: start_thread (pthread_create.c:477)
> ==22717==    by 0x4D5E132: clone (clone.S:95)
> ==22717==  If you believe this happened as a result of a stack
> ==22717==  overflow in your program's main thread (unlikely but
> ==22717==  possible), you can try to increase the size of the
> ==22717==  main thread stack using the --main-stacksize= flag.
> ==22717==  The main thread stack size used in this run was 8388608.
> ==22717== 
> ==22717== HEAP SUMMARY:
> ==22717==     in use at exit: 818,206,294 bytes in 332 blocks
> ==22717==   total heap usage: 707 allocs, 375 frees, 1,628,469,666 bytes allocated
> ==22717== 
> ==22717== Thread 1:
> ==22717== 1,520 bytes in 5 blocks are possibly lost in loss record 40 of 54
> ==22717==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
> ==22717==    by 0x40149DA: allocate_dtv (dl-tls.c:286)
> ==22717==    by 0x40149DA: _dl_allocate_tls (dl-tls.c:532)
> ==22717==    by 0x48AB322: allocate_stack (allocatestack.c:622)
> ==22717==    by 0x48AB322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
> ==22717==    by 0x487AECA: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
> ==22717==    by 0x48728E0: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
> ==22717==    by 0x5E4EF6: ncnn::UnaryOp_x86_fma::forward_inplace(ncnn::Mat&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x194034: ncnn::NetPrivate::do_forward_layer(ncnn::Layer const*, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x194FB8: ncnn::NetPrivate::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x194F87: ncnn::NetPrivate::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x194F87: ncnn::NetPrivate::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x194F87: ncnn::NetPrivate::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717==    by 0x194F87: ncnn::NetPrivate::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const (in /root/volume/ncnn/build/examples/test_llama_module)
> ==22717== 
> ==22717== LEAK SUMMARY:
> ==22717==    definitely lost: 0 bytes in 0 blocks
> ==22717==    indirectly lost: 0 bytes in 0 blocks
> ==22717==      possibly lost: 1,520 bytes in 5 blocks
> ==22717==    still reachable: 818,204,774 bytes in 327 blocks
> ==22717==                       of which reachable via heuristic:
> ==22717==                         multipleinheritance: 20,256 bytes in 65 blocks
> ==22717==         suppressed: 0 bytes in 0 blocks
> ==22717== Reachable blocks (those to which a pointer was found) are not shown.
> ==22717== To see them, rerun with: --leak-check=full --show-leak-kinds=all
> ==22717== 
> ==22717== For lists of detected and suppressed errors, rerun with: -s
> ==22717== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
> Segmentation fault (core dumped)

`

**Comment by parrotsky** [comment]
拿x86为例子， 增加对于三维tensor输入的判断， 将其折叠为二维调用。
```
    if (bottom_blob.dims == 2 && bottom_blob.w == num_input)
    {
        // gemm
        int h = bottom_blob.h;
        size_t elemsize = bottom_blob.elemsize;
        int elempack = bottom_blob.elempack;

        top_blob.create(num_output, h, elemsize, elempack, opt.blob_allocator);
        if (top_blob.empty())
            return -100;

        innerproduct_gemm_sse(bottom_blob, top_blob, weight_data_tm, bias_data, activation_type, activation_params, opt);

        return 0;
    }



    if (bottom_blob.dims == 3 && bottom_blob.c == num_input)
    {
        // 3d input vector 
        int w = bottom_blob.w;
        int h = bottom_blob.h;
        int c = bottom_blob.c; // num_input
        size_t elemsize = bottom_blob.elemsize;
        int elempack = bottom_blob.elempack;
 

        ncnn::Mat bottom_blob_flattened = bottom_blob.reshape(c, w*h );
        ncnn::Mat top_blob_flattened;

        // Adjust the size of top_blob
        top_blob.create(num_output, h * w, elemsize, elempack, opt.blob_allocator);
        if (top_blob.empty())
        return -100;

        // Perform the matrix multiplication
        innerproduct_gemm_sse(bottom_blob_flattened, top_blob, weight_data_tm, bias_data, activation_type, activation_params, opt);
        top_blob = top_blob.reshape(w, h, num_output);
         return 0;
    }

```
修改第5个linear前面的reshape参数，就不会出现内存问题。
Reshape                  reshape_10               1 1 47 48 0=1 1=128 2=4096
InnerProduct             linear_5                 1 1 48 49 0=4096 1=0 2=16777216
